File: java/test/OrganizationEditRoleEntityTest.java
package voxgig.hook0sdk.sdktest;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import org.junit.jupiter.api.Assumptions;
import org.junit.jupiter.api.Test;

import voxgig.hook0sdk.core.Helpers;
import voxgig.hook0sdk.core.SdkEntity;
import voxgig.hook0sdk.core.Hook0SDK;
import voxgig.hook0sdk.utility.Json;
import voxgig.hook0sdk.utility.struct.Struct;

@SuppressWarnings({"unchecked", "unused"})
public class OrganizationEditRoleEntityTest {

  @Test
  public void instance() {
    Hook0SDK testsdk = Hook0SDK.testSDK();
    SdkEntity ent = testsdk.organizationEditRole(null);
    assertNotNull(ent, "expected non-null organization_edit_role entity");
  }

  @Test
  public void basic() {
    RunnerSupport.EntityTestSetup setup = organizationEditRoleBasicSetup(null);
    // Per-op sdk-test-control.json skip — basic test exercises a flow
    // with multiple ops; skipping any op skips the whole flow.
    String mode = setup.live ? "live" : "unit";
    for (String op : new String[] { "update" }) {
      String reason = RunnerSupport.skipReason("entityOp", "organization_edit_role." + op, mode);
      Assumptions.assumeTrue(reason == null,
          reason == null || "".equals(reason)
              ? "skipped via sdk-test-control.json" : reason);
    }
    // The basic flow consumes synthetic IDs from the fixture. In live mode
    // without an *_ENTID env override, those IDs hit the live API and 4xx.
    Assumptions.assumeFalse(setup.syntheticOnly,
        "live entity test uses synthetic IDs from fixture — set HOOK0_TEST_ORGANIZATION_EDIT_ROLE_ENTID JSON to run live");
    Hook0SDK client = setup.client;

    // Bootstrap entity data from existing test data (no create step in flow).
    List<List<Object>> organizationEditRoleRef01DataRaw = Struct.items(Helpers.toMapAny(
        Struct.getpath(setup.data, "existing.organization_edit_role")));
    Map<String, Object> organizationEditRoleRef01Data = organizationEditRoleRef01DataRaw.isEmpty()
        ? null : Helpers.toMapAny(organizationEditRoleRef01DataRaw.get(0).get(1));

    // UPDATE
    SdkEntity organizationEditRoleRef01Ent = client.organizationEditRole(null);
    Map<String, Object> organizationEditRoleRef01DataUp0Up = new LinkedHashMap<>();

    String organizationEditRoleRef01MarkdefUp0Name = "role";
    String organizationEditRoleRef01MarkdefUp0Value = "Mark01-organization_edit_role_ref01_" + setup.now;
    organizationEditRoleRef01DataUp0Up.put(organizationEditRoleRef01MarkdefUp0Name, organizationEditRoleRef01MarkdefUp0Value);

    Object organizationEditRoleRef01ResdataUp0Result = organizationEditRoleRef01Ent.update(organizationEditRoleRef01DataUp0Up, null);
    Map<String, Object> organizationEditRoleRef01ResdataUp0 = Helpers.toMapAny(organizationEditRoleRef01ResdataUp0Result instanceof SdkEntity ? ((SdkEntity) organizationEditRoleRef01ResdataUp0Result).data() : organizationEditRoleRef01ResdataUp0Result);
    assertNotNull(organizationEditRoleRef01ResdataUp0, "expected update result to be a map");
    assertEquals(organizationEditRoleRef01MarkdefUp0Value, organizationEditRoleRef01ResdataUp0.get(organizationEditRoleRef01MarkdefUp0Name),
        "expected " + organizationEditRoleRef01MarkdefUp0Name + " to be updated");

  }

  static RunnerSupport.EntityTestSetup organizationEditRoleBasicSetup(Map<String, Object> extra) {
    RunnerSupport.loadEnvLocal();

    Map<String, Object> entityData;
    try {
      String entityDataSource = Files.readString(Path.of(
          "..", ".sdk", "test", "entity", "organization_edit_role", "OrganizationEditRoleTestData.json"));
      entityData = Helpers.toMapAny(Json.parse(entityDataSource));
    }
    catch (Exception e) {
      throw new AssertionError("failed to read organization_edit_role test data: " + e.getMessage(), e);
    }

    Map<String, Object> options = new LinkedHashMap<>();
    options.put("entity", entityData.get("existing"));

    Hook0SDK client = Hook0SDK.testSDK(options, extra);

    // Generate idmap via transform, matching TS pattern.
    List<Object> idnames = new ArrayList<>();
    idnames.add("organization_edit_role01");
    idnames.add("organization_edit_role02");
    idnames.add("organization_edit_role03");
    Object idmap = Struct.transform(idnames, Json.parse(
        "{\"`$PACK`\": [\"\", {"
        + "\"`$KEY`\": \"`$COPY`\","
        + "\"`$VAL`\": [\"`$FORMAT`\", \"upper\", \"`$COPY`\"]"
        + "}]}"));

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against
    // synthetic IDs from the fixture and 4xx's. Surface this so the test
    // can skip.
    String entidEnvRaw = RunnerSupport.getenv("HOOK0_TEST_ORGANIZATION_EDIT_ROLE_ENTID");
    boolean idmapOverridden = entidEnvRaw != null
        && entidEnvRaw.trim().startsWith("{");

    Map<String, Object> envm = new LinkedHashMap<>();
    envm.put("HOOK0_TEST_ORGANIZATION_EDIT_ROLE_ENTID", idmap);
    envm.put("HOOK0_TEST_LIVE", "FALSE");
    envm.put("HOOK0_TEST_EXPLAIN", "FALSE");
    envm.put("HOOK0_APIKEY", "NONE");
    Map<String, Object> env = RunnerSupport.envOverride(envm);

    Map<String, Object> idmapResolved = Helpers.toMapAny(env.get("HOOK0_TEST_ORGANIZATION_EDIT_ROLE_ENTID"));
    if (idmapResolved == null) {
      idmapResolved = Helpers.toMapAny(idmap);
    }

    boolean live = "TRUE".equals(env.get("HOOK0_TEST_LIVE"));
    if (live) {
      Map<String, Object> liveOpts = new LinkedHashMap<>();
      liveOpts.put("apikey", env.get("HOOK0_APIKEY"));
      Object mergedOpts = Struct.merge(Struct.jt(liveOpts, extra));
      client = new Hook0SDK(Helpers.toMapAny(mergedOpts));
    }

    RunnerSupport.EntityTestSetup setup = new RunnerSupport.EntityTestSetup();
    setup.client = client;
    setup.data = entityData;
    setup.idmap = idmapResolved;
    setup.env = env;
    setup.explain = "TRUE".equals(env.get("HOOK0_TEST_EXPLAIN"));
    setup.live = live;
    setup.syntheticOnly = live && !idmapOverridden;
    setup.now = System.currentTimeMillis();
    return setup;
  }
}

File: php/test/OrganizationEditRoleEntityTest.php
<?php
declare(strict_types=1);

// OrganizationEditRole entity test

require_once __DIR__ . '/../hook0_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class OrganizationEditRoleEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = Hook0SDK::test(null, null);
        $ent = $testsdk->OrganizationEditRole(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = organization_edit_role_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["update"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "organization_edit_role." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set HOOK0_TEST_ORGANIZATION_EDIT_ROLE_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // Bootstrap entity data from existing test data.
        $organization_edit_role_ref01_data_raw = Vs::items(Helpers::to_map(
            Vs::getpath($setup["data"], "existing.organization_edit_role")));
        $organization_edit_role_ref01_data = null;
        if (count($organization_edit_role_ref01_data_raw) > 0) {
            $organization_edit_role_ref01_data = Helpers::to_map($organization_edit_role_ref01_data_raw[0][1]);
        }

        // UPDATE
        $organization_edit_role_ref01_ent = $client->OrganizationEditRole(null);
        $organization_edit_role_ref01_data_up0_up = [
        ];

        $organization_edit_role_ref01_markdef_up0_name = "role";
        $organization_edit_role_ref01_markdef_up0_value = "Mark01-organization_edit_role_ref01_" . $setup["now"];
        $organization_edit_role_ref01_data_up0_up[$organization_edit_role_ref01_markdef_up0_name] = $organization_edit_role_ref01_markdef_up0_value;

        $organization_edit_role_ref01_resdata_up0_result = $organization_edit_role_ref01_ent->update($organization_edit_role_ref01_data_up0_up, null);
        $organization_edit_role_ref01_resdata_up0 = Helpers::to_map(is_object($organization_edit_role_ref01_resdata_up0_result) && method_exists($organization_edit_role_ref01_resdata_up0_result, 'data_get') ? $organization_edit_role_ref01_resdata_up0_result->data_get() : $organization_edit_role_ref01_resdata_up0_result);
        $this->assertNotNull($organization_edit_role_ref01_resdata_up0);
        $this->assertEquals($organization_edit_role_ref01_resdata_up0[$organization_edit_role_ref01_markdef_up0_name], $organization_edit_role_ref01_markdef_up0_value);

    }
}

function organization_edit_role_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/organization_edit_role/OrganizationEditRoleTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = Hook0SDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["organization_edit_role01", "organization_edit_role02", "organization_edit_role03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("HOOK0_TEST_ORGANIZATION_EDIT_ROLE_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "HOOK0_TEST_ORGANIZATION_EDIT_ROLE_ENTID" => $idmap,
        "HOOK0_TEST_LIVE" => "FALSE",
        "HOOK0_TEST_EXPLAIN" => "FALSE",
        "HOOK0_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["HOOK0_TEST_ORGANIZATION_EDIT_ROLE_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["HOOK0_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["HOOK0_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new Hook0SDK(Helpers::to_map($merged_opts));
    }

    $live = $env["HOOK0_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["HOOK0_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
